src/Lecture2.hs

```diff
@@ -313,8 +319,17 @@ mergeSort :: [Int] -> [Int]
 mergeSort [] = []
 mergeSort [x] = [x]
 mergeSort xs = merge (mergeSort first) (mergeSort second)
-    where (first, second) = split xs
-          split xs' = splitAt (length xs' `div` 2) xs'
+    where (first, second) = separate xs
+
+-- Original implementation:
+--     where (first, second) = split xs
+--          split xs' = splitAt (length xs' `div` 2) xs'
+
+separate :: [Int] -> ([Int], [Int])
+separate xs = go ([], []) xs
+    where go (first, second) [] = (reverse first, reverse second)
+          go (first, second) (x:[]) = (reverse $ x:first, reverse $ second)
+          go (first, second) (x:y:xs') = go (x:first, y:second) xs'
```

Comment:
`reverse` still requires a separate list traversal and it won't be more efficient. To Split and transverse the list in a single pass, you can do something like this:

```haskell
separate [] = ([], [])
separate [x] = ([x], [])
separate (x : y : zs) =
    let (xs, ys) = separate zs
    in (x : xs, y : ys)
```

(it'll require one more advanced feature to actually make it fast but that's out of the scope for this course 🙃 )